manager/allocator/network.go

```diff
@@ -682,179 +731,188 @@ func (a *Allocator) allocateNetwork(ctx context.Context, n *api.Network) error {
 	nc := a.netCtx
+	var allocatedServices []*api.Service
 	for _, s := range nc.unallocatedServices {
 		if !nc.nwkAllocator.IsServiceAllocated(s) {
 			if err := a.allocateService(ctx, s); err != nil {
-				log.G(ctx).Debugf("Failed allocation of unallocated service %s: %v", s.ID, err)
+				log.G(ctx).WithError(err).Debugf("Failed allocation of unallocated service %s", s.ID)
 				continue
 			}
+			allocatedServices = append(allocatedServices, s)
 		}
-
-		delete(nc.unallocatedServices, s.ID)
 	}
-}
-
-func (a *Allocator) procUnallocatedTasksNetwork(ctx context.Context) {
-	nc := a.netCtx
-	tasks := make([]*api.Task, 0, len(nc.unallocatedTasks))
 
-	committed, err := a.store.Batch(func(batch *store.Batch) error {
-		for _, t := range nc.unallocatedTasks {
-			var allocatedT *api.Task
-			err := batch.Update(func(tx store.Tx) error {
-				var err error
-				allocatedT, err = a.allocateTask(ctx, tx, t)
-				return err
-			})
+	if len(allocatedServices) == 0 {
+		return
+	}
 
-			if err != nil {
-				log.G(ctx).WithError(err).Error("task allocation failure")
+	if _, err := a.store.Batch(func(batch *store.Batch) error {
+		for _, s := range allocatedServices {
+			if err := a.commitAllocatedService(ctx, batch, s); err != nil {
+				log.G(ctx).WithError(err).Debugf("Failed to commit allocation of unallocated service %s", s.ID)
 				continue
 			}
 
-			tasks = append(tasks, allocatedT)
+			delete(nc.unallocatedServices, s.ID)
 		}
-
 		return nil
-	})
-
-	if err != nil {
-		log.G(ctx).WithError(err).Error("failed a store batch operation while processing unallocated tasks")
+	}); err != nil {
+		log.G(ctx).WithError(err).Error("Failed to commit allocation of unallocated services")
 	}
+}
 
-	var retryCnt int
-	for len(tasks) != 0 {
-		var err error
-
-		for _, t := range tasks[:committed] {
-			delete(nc.unallocatedTasks, t.ID)
-		}
+func (a *Allocator) procUnallocatedTasksNetwork(ctx context.Context) {
+	nc := a.netCtx
+	allocatedTasks := make([]*api.Task, 0, len(nc.unallocatedTasks))
 
-		tasks = tasks[committed:]
-		if len(tasks) == 0 {
-			break
+	for _, t := range nc.unallocatedTasks {
+		if err := a.allocateTask(ctx, t); err == nil {
+			allocatedTasks = append(allocatedTasks, t)
+		} else if err != errNoChanges {
+			log.G(ctx).WithError(err).Error("task allocation failure")
 		}
+	}
 
-		updatedTasks := make([]*api.Task, 0, len(tasks))
-		committed, err = a.store.Batch(func(batch *store.Batch) error {
-			for _, t := range tasks {
-				err := batch.Update(func(tx store.Tx) error {
-					return store.UpdateTask(tx, t)
-				})
+	if len(allocatedTasks) == 0 {
+		return
+	}
 
-				if err != nil {
-					log.G(ctx).WithError(err).Error("allocated task store update failure")
-					continue
-				}
+	if _, err := a.store.Batch(func(batch *store.Batch) error {
+		for _, t := range allocatedTasks {
+			err := batch.Update(func(tx store.Tx) error {
+				return a.commitAllocatedTask(ctx, batch, t)
```

Comment:
I couldn't find out where we are removing the `task` from `nc.unallocatedTasks` after a successful commit to the store. If it is not done, we should do it.